### PR TITLE
qpdf: split outputs; modernize

### PR DESCRIPTION
**Pull request by getchoo**
## Description of changes

Came across `lib.getExe` not working with this package in https://github.com/NixOS/nixpkgs/pull/326372

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by getchoo** [comment]
rebased onto staging as this will cause mass rebuilds

**Comment by doronbehar** [comment]
@ofborg eval

File: pkgs/development/libraries/qpdf/default.nix [hunk 1]
```diff
@@ -47,6 +48,8 @@ stdenv.mkDerivation rec {
   doCheck = true;
 
   passthru.tests = {
+    pkg-config = testers.hasPkgConfigModules { package = finalAttrs.finalPackage; };
+    version = testers.testVersion { package = finalAttrs.finalPackage; };
```
Comment (doronbehar) [inline pkgs/development/libraries/qpdf/default.nix:52]: I'd vote for using the new `versionCheckHook` if possible.

**Review by doronbehar — COMMENTED** [review]
Looks good overall!

File: pkgs/development/libraries/qpdf/default.nix (hunk 1 above)
Comment (getchoo) [reply pkgs/development/libraries/qpdf/default.nix]: thanks! i hadn't come across that yet

File: pkgs/development/libraries/qpdf/default.nix [hunk 2]
```diff
@@ -11,20 +11,31 @@
...
+  outputs = [
+    "bin"
+    "dev"
+    "doc"
+    "lib"
+    "man"
+    "out"
+  ];
+
+  nativeBuildInputs = [ cmake perl versionCheckHook ];
```
Comment (doronbehar) [inline pkgs/development/libraries/qpdf/default.nix:38]: If [the docs for versionCheckHook](https://github.com/NixOS/nixpkgs/blob/7e7c39ea35c5cdd002cd4588b03a3fb9ece6fad9/doc/hooks/versionCheckHook.section.md?plain=1#L16-L19) are correct, then you should do:

```suggestion
  nativeBuildInputs = [
    cmake
    perl
  ];
  
  nativeInstallCheckInputs = [
    versionCheckHook
  ];
  doInstallCheck = true;

```

File: pkgs/development/libraries/qpdf/default.nix (hunk 1 above)
Comment (doronbehar) [reply pkgs/development/libraries/qpdf/default.nix]: > thanks! i hadn't come across that yet

I'm glad you liked it! (I wrote it)

File: pkgs/development/libraries/qpdf/default.nix (hunk 2 above)
Comment (getchoo) [reply pkgs/development/libraries/qpdf/default.nix]: I remember reading this when you first reviewed but then I completely forgot 🤦🏻
Fixed!

**Comment by doronbehar** [comment]
Splitting outputs sometimes causes breakages in dependent packages... I tested the build locally, (against `nixos-unstable`), and the contents in the different outputs seems almost perfect to me. I'll list it here just for reference:

```
$ tree $(nix build --print-out-paths --no-link -f. qpdf.{bin,dev,doc,lib,man,out}
/nix/store/3c9a59g8gcfpjsibp4lcx0r6rw7xkvq6-qpdf-11.9.1-bin
└── bin
    ├── fix-qdf
    ├── qpdf
    └── zlib-flate
/nix/store/vbyg26aw7xshfdhwn1h3c6v6j1hhzhyk-qpdf-11.9.1-dev
├── include
│   └── qpdf
│       ├── auto_job_c_att.hh
│       ├── auto_job_c_copy_att.hh
│       ├── auto_job_c_enc.hh
│       ├── auto_job_c_main.hh
│       ├── auto_job_c_pages.hh
│       ├── auto_job_c_set_page_labels.hh
│       ├── auto_job_c_uo.hh
│       ├── Buffer.hh
│       ├── BufferInputSource.hh
│       ├── ClosedFileInputSource.hh
│       ├── Constants.h
│       ├── DLL.h
│       ├── FileInputSource.hh
│       ├── InputSource.hh
│       ├── JSON.hh
│       ├── PDFVersion.hh
│       ├── Pipeline.hh
│       ├── Pl_Buffer.hh
│       ├── Pl_Concatenate.hh
│       ├── Pl_Count.hh
│       ├── Pl_DCT.hh
│       ├── Pl_Discard.hh
│       ├── Pl_Flate.hh
│       ├── Pl_Function.hh
│       ├── Pl_OStream.hh
│       ├── Pl_QPDFTokenizer.hh
│       ├── Pl_RunLength.hh
│       ├── Pl_StdioFile.hh
│       ├── Pl_String.hh
│       ├── PointerHolder.hh
│       ├── QIntC.hh
│       ├── QPDFAcroFormDocumentHelper.hh
│       ├── QPDFAnnotationObjectHelper.hh
│       ├── qpdf-c.h
│       ├── QPDFCryptoImpl.hh
│       ├── QPDFCryptoProvider.hh
│       ├── QPDFDocumentHelper.hh
│       ├── QPDFEFStreamObjectHelper.hh
│       ├── QPDFEmbeddedFileDocumentHelper.hh
│       ├── QPDFExc.hh
│       ├── QPDFFileSpecObjectHelper.hh
│       ├── QPDFFormFieldObjectHelper.hh
│       ├── QPDF.hh
│       ├── qpdfjob-c.h
│       ├── QPDFJob.hh
│       ├── qpdflogger-c.h
│       ├── QPDFLogger.hh
│       ├── QPDFMatrix.hh
│       ├── QPDFNameTreeObjectHelper.hh
│       ├── QPDFNumberTreeObjectHelper.hh
│       ├── QPDFObjectHandle.hh
│       ├── QPDFObjectHelper.hh
│       ├── QPDFObject.hh
│       ├── QPDFObjGen.hh
│       ├── QPDFOutlineDocumentHelper.hh
│       ├── QPDFOutlineObjectHelper.hh
│       ├── QPDFPageDocumentHelper.hh
│       ├── QPDFPageLabelDocumentHelper.hh
│       ├── QPDFPageObjectHelper.hh
│       ├── QPDFStreamFilter.hh
│       ├── QPDFSystemError.hh
│       ├── QPDFTokenizer.hh
│       ├── QPDFUsage.hh
│       ├── QPDFWriter.hh
│       ├── QPDFXRefEntry.hh
│       ├── QTC.hh
│       ├── QUtil.hh
│       ├── RandomDataProvider.hh
│       └── Types.h
├── lib
│   ├── cmake
│   │   └── qpdf
│   │       ├── libqpdfTargets.cmake
│   │       ├── libqpdfTargets-release.cmake
│   │       ├── qpdfConfig.cmake
│   │       └── qpdfConfigVersion.cmake
│   └── pkgconfig
│       └── libqpdf.pc
└── nix-support
    └── propagated-build-inputs
/nix/store/4zhsaqmd06s4fcgmpdjg7a23xhx15j2j-qpdf-11.9.1-doc
└── share
    └── doc
        └── qpdf
            ├── examples
            │   ├── extend-c-api.c
            │   ├── extend-c-api.h
            │   ├── extend-c-api-impl.cc
            │   ├── pdf-attach-file.cc
            │   ├── pdf-bookmarks.cc
            │   ├── pdf-c-objects.c
            │   ├── pdf-count-strings.cc
            │   ├── pdf-create.cc
            │   ├── pdf-custom-filter.cc
            │   ├── pdf-double-page-size.cc
            │   ├── pdf-filter-tokens.cc
            │   ├── pdf-invert-images.cc
            │   ├── pdf-linearize.c
            │   ├── pdf-mod-info.cc
            │   ├── pdf-name-number-tree.cc
            │   ├── pdf-npages.cc
            │   ├── pdf-overlay-page.cc
            │   ├── pdf-parse-content.cc
            │   ├── pdf-set-form-values.cc
            │   ├── pdf-split-pages.cc
            │   ├── qpdf-job.cc
            │   ├── qpdfjob-c.c
            │   ├── qpdfjob-c-save-attachment.c
            │   ├── qpdfjob-remove-annotations.cc
            │   └── qpdfjob-save-attachment.cc
            └── README-doc.txt
/nix/store/yzn4h9lfrscfkqjmzs6aw35g7qr2lw3g-qpdf-11.9.1-lib
└── lib
    ├── libqpdf.a
    ├── libqpdf.so -> libqpdf.so.29
    ├── libqpdf.so.29 -> libqpdf.so.29.9.1
    └── libqpdf.so.29.9.1
/nix/store/1h14r7rfxgr2ihgwani0zsq66zdsrqfn-qpdf-11.9.1-man
└── share
    └── man
        └── man1
            ├── fix-qdf.1.gz
            ├── qpdf.1.gz
            └── zlib-flate.1.gz
/nix/store/rcz5hfbl23hg02330dwff1isbw6haicf-qpdf-11.9.1

21 directories, 111 files
```

The only problem I found, was the contents of `${qpdf.dev}/lib/cmake/qpdf/libqpdfTargets.cmake`:

```cmake
# The installation prefix configured by this project.
set(_IMPORT_PREFIX "/nix/store/rcz5hfbl23hg02330dwff1isbw6haicf-qpdf-11.9.1")

# Create imported target qpdf::libqpdf
add_library(qpdf::libqpdf SHARED IMPORTED)

set_target_properties(qpdf::libqpdf PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)

# Create imported target qpdf::libqpdf_static
add_library(qpdf::libqpdf_static STATIC IMPORTED)

set_target_properties(qpdf::libqpdf_static PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "/nix/store/rc18a8k50zmrif61250sfidkqvlg41ln-zlib-1.3.1/lib/libz.so;/nix/store/0dsiifpjc1cxbaa8d6b0kc6d2wijk7in-libjpeg-turbo-3.0.3/lib/libjpeg.so"
)
```

Note how the `out` output is an empty directory, and the `_IMPORT_PREFIX` is pointing to it... Perhaps the `.dev` output should be removed, and then the `out` output will include the cmake, pkgconfig files etc, and the cmake files will use the correct prefix.

**Comment by getchoo** [comment]
> Note how the out output is an empty directory, and the _IMPORT_PREFIX is pointing to it... Perhaps the .dev output should be removed, and then the out output will include the cmake, pkgconfig files etc, and the cmake files will use the correct prefix.

Thanks for catching this! I've built a few packages with this change as well and all seems good :)

**Comment by doronbehar** [comment]
> > Note how the out output is an empty directory, and the _IMPORT_PREFIX is pointing to it... Perhaps the .dev output should be removed, and then the out output will include the cmake, pkgconfig files etc, and the cmake files will use the correct prefix.
> 
> Thanks for catching this! I've built a few packages with this change as well and all seems good :)

I imagine that the packages you tested were using the pkgconfig file, and not the cmake API. However, inspecting in a similar manner the outputs' contents shows promising prospect.

**Comment by doronbehar** [comment]
I will merge once ofborg evaluation is finished.

**Pull request merged**